sylt-compiler/src/name_resolution.rs

```diff
@@ -0,0 +1,1038 @@
+                    self.push_var(n.clone(), VarKind::Const);
+                }
+                let ret = self.ty(ret)?;
+                let body = self.block(body)?;
+                self.stack.truncate(ss);
+                E::Function { name, params, ret, body, pure: *pure, span }
+            }
+            EK::Blob { blob, fields: parser_fields } => {
+                let blob = self.ty_assignable(blob)?;
+                let mut fields = Vec::new();
+                for (name, field) in parser_fields.iter() {
+                    fields.push((name.clone(), self.expression(field)?));
+                }
+                E::Blob { blob, fields, span }
+            }
+            EK::Tuple(values) => self.collection(Collection::Tuple, &values, span)?,
+            EK::List(values) => self.collection(Collection::List, &values, span)?,
+            EK::Set(values) => self.collection(Collection::Set, &values, span)?,
+            EK::Dict(values) => self.collection(Collection::Dict, &values, span)?,
+            EK::Float(f) => E::Float(*f, span),
+            EK::Int(i) => E::Int(*i, span),
+            EK::Str(s) => E::Str(s.clone(), span),
+            EK::Bool(b) => E::Bool(*b, span),
+            EK::Nil => E::Nil(span),
+        })
+    }
+
+    fn statement(&mut self, stmt: &ParserStatement) -> ResolveResult<Option<Statement>> {
+        use sylt_parser::StatementKind as SK;
+        use Statement as S;
+        let span = stmt.span;
+        Ok(match &stmt.kind {
+            // These are already handled
+            SK::Blob { .. }
+            | SK::EmptyStatement
+            | SK::Enum { .. }
+            | SK::FromUse { .. }
+            | SK::Use { .. } => None,
+
+            SK::ExternalDefinition { ident, kind, ty } => Some(S::ExternalDefinition {
+                ident: ident.clone(),
+                kind: *kind,
+                span: ident.span,
+                ty: self.ty(ty)?,
+            }),
+
+            SK::Definition { ident, kind, ty, value } => {
+                let ss = self.stack.len();
+                let value = match (
+                    matches!(value.kind, sylt_parser::ExpressionKind::Function { .. }),
+                    ss == 0,
+                ) {
+                    (false, true) => {
+                        // Outer statement - it's a global so just evaluate the value!
+                        self.expression(value)?
+                    }
+                    (true, _) => {
+                        // Function, push the var before!
+                        self.push_var(ident.clone(), *kind);
+                        self.expression(value)?
+                    }
+                    (false, false) => {
+                        // Value, push the var after!
+                        let expr = self.expression(value)?;
+                        self.push_var(ident.clone(), *kind);
+                        expr
+                    }
+                };
```

Comment:
Kind of a confusing match on `(bool, bool)` imo.
```suggestion
                let value = if matches!(
                    value.kind,
                    sylt_parser::ExpressionKind::Function { .. },
                ) {
                    // Function, push the var before!
                    self.push_var(ident.clone(), *kind);
                    self.expression(value)?
                } else if self.stack.is_empty() {
                    // Outer statement - it's a global so just evaluate the value!
                    self.expression(value)?
                } else {
                    // Value, push the var after!
                    let expr = self.expression(value)?;
                    self.push_var(ident.clone(), *kind);
                    expr
                };
```

Reply:
Also, it looks like global functions get pushed even though they should already be handled in a previous pass? Is this intentional?

Reply:
This is the only piece of complicated logic in the program, and your translation won't match the behaviour of my match-block. And I've recently convinced myself my matchblock is wrong and your if-statements are incorrect. But it might be more readable as an if-else chain?

And you're point that `global functions get pushed` are correct, this is not the behaviour we want. We want local variables to live on the stack. Or maybe I've thought about this, because we need to push something to show that we're inside a function. Maybe we should push a dummy value for entering a function - or setting a flag. At this point I'm rambling. But yes, this is wrong.